flocker/volume/filesystems/zfs.py

```diff
@@ -289,33 +341,46 @@ def writer(self):
             # a hack.  When we replace this mechanism with a proper API we
             # should make it include that information.
             #
-            # -e means "if the stream says it is for foo/bar/baz then receive
-            # into baz".  I don't know why self.name is also required,
-            # then. XXX try -d self.pool instead. XXX it works without -e w/
-            # self.name too. XXX Delete this paragraph if we go ahead with just
-            # `-F` in the implementation.
-            #
-            # -F means force.  If the stream is based on not-quite-the-latest
+            # If the stream is based on not-quite-the-latest
             # snapshot then we have to throw away all the snapshots newer than
             # it in order to receive the stream.  To do that you have to
             # force.
             #
-            cmd = [b"zfs", b"receive", b"-F", self.name]
+            force = True
         else:
             # If the filesystem doesn't already exist then this is a complete
             # data stream.
-            cmd = [b"zfs", b"receive", self.name]
-        process = Popen(cmd, stdin=PIPE)
-        succeeded = False
+            force = False
+
+        (rfd, wfd) = os.pipe()
+        wfile = os.fdopen(wfd, "w")
+        queue = Queue()
+
+        def recv_and_close():
+            try:
+                libzfs_core.lzc_receive(self.name, rfd, force)
+                success = True
+            except Exception:
+                success = False
+            finally:
+                os.close(rfd)
+                queue.put(success)
+
+        d = self._async_lzc.callDeferred(recv_and_close)
+        d.addBoth(lambda _: None)
```

Comment:
This seems to be the same as the reader / sender method above in that it uses the Twisted threadpool but then blocks until the thread has completed rather than returning the deferred.
Perhaps I'm missing some context. Perhaps we're working toward a fully async implementation of the ZFS backend in...but in steps. If so it'd be useful to read a few docstrings and comments explaining the plan.